File: golang/gate/gate_test.go
package gate

import (
	"testing"

	"github.com/muthuishere/citenexus/golang/internal/conform"
)

// expectedFaithfulCounts pins the bucket sizes of conformance/cases/faithful.json.
// Iterating every case is only a contract while the case list cannot silently
// shrink underneath it.
var expectedFaithfulCounts = map[string]int{
	"supported": 7,
	"relevance": 5,
}

type faithfulFixture struct {
	Supported []struct {
		Answer    string `json:"answer"`
		Passage   string `json:"passage"`
		Supported bool   `json:"supported"`
	} `json:"supported"`
	Relevance []struct {
		Query    string `json:"query"`
		Passage  string `json:"passage"`
		Relevant bool   `json:"relevant"`
	} `json:"relevance"`
}

func loadFaithful(t *testing.T) faithfulFixture {
	t.Helper()
	var fixture faithfulFixture
	conform.Case(t, "faithful.json", &fixture)
	return fixture
}

func assertFaithfulCounts(t *testing.T, fixture faithfulFixture) {
	t.Helper()
	for name, got := range map[string]int{
		"supported": len(fixture.Supported),
		"relevance": len(fixture.Relevance),
	} {
		if want := expectedFaithfulCounts[name]; got != want {
			t.Fatalf("faithful.json bucket %q: got %d vectors, want %d", name, got, want)
		}
	}
}

func TestFaithfulVectorBucketSizes(t *testing.T) {
	assertFaithfulCounts(t, loadFaithful(t))
}

// The gates are proven against the shared fixture — every "supported" and
// "relevance" case must match the Python reference (citenexus.answer.verify)
// exactly. Follows the tokenize exemplar: load the fixture, assert over ALL
// cases, no leniency.
func TestGateConformance(t *testing.T) {
	fixture := loadFaithful(t)
	assertFaithfulCounts(t, fixture)

	for _, c := range fixture.Supported {
		if got := IsSupported(c.Answer, c.Passage); got != c.Supported {
			t.Errorf("IsSupported(%q, %q) = %v, want %v", c.Answer, c.Passage, got, c.Supported)
		}
	}

	for _, c := range fixture.Relevance {
		if got := HasRelevanceOverlap(c.Query, c.Passage); got != c.Relevant {
			t.Errorf("HasRelevanceOverlap(%q, %q) = %v, want %v", c.Query, c.Passage, got, c.Relevant)
		}
	}
}

File: js/src/gate/gate.test.ts
import { describe, it, expect } from "vitest";
import { loadCase } from "../conform/fixtures.js";
import { hasRelevanceOverlap, isSupported } from "./gate.js";

// The gate is proven against the shared fixture — every case must match the
// Python reference (citenexus.answer.verify) exactly. Follows the tokenize
// exemplar: load conformance/cases/faithful.json, assert over ALL cases, no
// leniency.
interface Faithful {
  supported: { answer: string; passage: string; supported: boolean }[];
  relevance: { query: string; passage: string; relevant: boolean }[];
}

/** Bucket sizes, pinned. A vector silently dropped from a bucket is a weakened
 *  contract that no per-case assertion can see. */
const EXPECTED_COUNTS: Record<string, number> = { supported: 7, relevance: 5 };

describe("gate conformance", () => {
  const fixture = loadCase<Faithful>("faithful.json");

  it("bucket names and sizes are pinned", () => {
    const sizes = Object.fromEntries(
      Object.entries(fixture).map(([k, v]) => [k, (v as unknown[]).length]),
    );
    expect(sizes).toEqual(EXPECTED_COUNTS);
  });

  for (const c of fixture.supported) {
    it(`isSupported(${JSON.stringify(c.answer)}, ${JSON.stringify(c.passage)})`, () => {
      expect(isSupported(c.answer, c.passage)).toBe(c.supported);
    });
  }

  for (const c of fixture.relevance) {
    it(`hasRelevanceOverlap(${JSON.stringify(c.query)}, ${JSON.stringify(c.passage)})`, () => {
      expect(hasRelevanceOverlap(c.query, c.passage)).toBe(c.relevant);
    });
  }
});
